src/Microsoft.OData.Core/Metadata/EdmLibraryExtensions.cs

```diff
@@ -305,13 +332,15 @@ internal static IEnumerable<IEdmOperation> FilterOperationsByParameterNames(this
 
                 // if any parameter count is different, don't consider it a match.
                 List<IEdmOperationParameter> functionImportParameters = parametersToMatch.ToList();
-                if (functionImportParameters.Count != parametersList.Count)
+
+                // if any required parameters are missing, don't consider it a match.
+                if (functionImportParameters.Where(p => !(p is IEdmOptionalParameter)).Any(p => parametersList.All(k => !string.Equals(k, p.Name, caseInsensitive ? StringComparison.OrdinalIgnoreCase : StringComparison.Ordinal))))
                 {
                     continue;
                 }
 
-                // if any parameter was missing, don't consider it a match.
-                if (functionImportParameters.Any(p => parametersList.All(k => !string.Equals(k, p.Name, caseInsensitive ? StringComparison.OrdinalIgnoreCase : StringComparison.Ordinal))))
+                // if any specified parameters don't match, don't consider it a match.
+                if (parametersList.Any(k => functionImportParameters.All(p => !string.Equals(k, p.Name, caseInsensitive ? StringComparison.OrdinalIgnoreCase : StringComparison.Ordinal))))

```

Comment:
Yes; as the TODO in FilterOperationsByParameterNames suggests, there is a great deal of redundancy between FilterOperationByParameterNames and FilterFunctionsByParameterNames. I didn't try and fix that with this PR, but let me take a look...

---
In reply to: [123059774](https://github.com/OData/odata.net/pull/828#discussion_r123059774) [](ancestors = 123059774)